Cover: Avoid adding a wrapper Group block when transforming to Group, where possible

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Depends on #40497 Check that a transform matches at the time of running the transform 

## What?
<!-- In a few words, what is the PR actually doing? -->

Following on from #40212, this PR seeks to fix / improve some of the behaviour when transforming from a Cover block to a Group block. 

The changes are:

* If the Cover block contains only a single Group block as a direct child, then remove the Cover block as the wrapper, and merge in the Cover block's attributes / overlay colors with the Group block's attributes, so that the background color is preserved.
* If the Cover block contains any other mix of blocks as direct children, then convert the Cover block directly to a Group block, with the overlay colors transformed to the appropriate background color attributes on the Group block.
* If the Cover block contained a `url` attribute (so, it had some form of media as a background), then retain the existing behaviour where we defer to the Group block's default transform that adds a Group block as a wrapper — this way the user doesn't unexpectedly lose content.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As noted in the review on that PR, when transforming from a Cover block to a Group block, the default transform in the Group block causes a Group block to be added as a wrapper. Repeatedly transforming between Group and Cover and back again results in repeatedly nesting the set of blocks in further levels of Group blocks. With this PR, in most cases, we should now be able to switch between Group -> Cover -> Group again without incurring additional nesting.

Note: this PR makes lots of assumptions about user intent — the _hope_ is that the changes here get the behaviour closer to what users might expect when transforming from Cover to Group, but it _is_ making assumptions, and the transform is still a destructive action, so it isn't perfect. Happy for feedback (or pushback) on the change if folks think there's a better way to go about this.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add a transform from Cover to Group to the Cover block
* In the transform, transpose overlay values to the appropriate background values
* Attempt to preserve values in the merged block in a way that's logical and reduces how destructive the transform is

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Create a post containing a mix of Cover blocks that contain various kinds of Group, Row, and Stack blocks as children (and other arbitrary blocks)
2. Try variations where the inner Group block has background or gradient colors, and other settings
3. Try transforming the Cover block to a Group block, and see if the result is what you expected
4. Try creating a Group block with a gradient or background color, transform to Cover, then transform back to Group — the block should be in the same(ish) state as it was originally

Tests are also included, and can be run from a terminal via:

```
npm run test-unit -- --testPathPattern packages/block-library/src/cover/test/transforms.js
```

## Screenshots or screencast <!-- if applicable -->

| Group -> Cover -> Group | Row -> Cover -> Row |
| --- | --- |
| ![2022-04-13 16 02 50](https://user-images.githubusercontent.com/14988353/163110923-411835c0-9603-4ab2-adfe-07b66783b97d.gif) | ![2022-04-13 16 09 01](https://user-images.githubusercontent.com/14988353/163111467-d98bb206-1558-4618-94a6-4a4cd7e3bd97.gif) |

andrewserong committed:
Cover: Avoid adding a wrapper Group block when transforming to Group, where possible

andrewserong committed:
Use cleanEmptyObject to ensure stray undefined values from the Group block don't override real values from the Cover block

andrewserong commented:
Note: I'll be AFK for the next couple of weeks, so I likely won't see any replies on this PR. If anyone would like to take this over and / or merge it while I'm away, feel free to! 🙇

Commit:
Only add the style attribute object if style attribute is set

Commit:
Stop empty style attribute being added when converting from group to cover

glendaviesnz commented:
This tested well for me, but added an empty `style: { color: {}}` attribute if custom colors were not set, so have pushed an couple of updates to prevent this.

aaronrobertshaw reviewed (changes requested):
Thanks for addressing this one @andrewserong 👍 

This tested well for me for transforming to and from Group blocks (including Row & Stack) with either flat color or gradient backgrounds.

> If the Cover block contained a url attribute (so, it had some form of media as a background), then retain the existing behaviour where we defer to the Group block's default transform that adds a Group block as a wrapper.


Unfortunately, for me if I select a background image for the cover block and then attempt to transform to a Group block, I lose the image rather than the entire Cover block being wrapped in a new Group block.


https://user-images.githubusercontent.com/60436221/163312419-d8ab5471-80aa-4225-b8f0-599c619bbdff.mp4

It appears that the new transform is still used despite the `isMatch` correctly returning false. It might also pay for us to update the [grouping related tests.](https://github.com/WordPress/gutenberg/blob/1168512a6fde5d2f8815d4b84d992e36871723aa/packages/e2e-tests/specs/editor/various/block-grouping.test.js) 

I'll dig into this in more detail next week.

ramonjd commented:
> Attempt to preserve values in the merged block in a way that's logical and reduces how destructive the transform is

Should custom background and gradient background colors be migrated as well? As far as I can tell they aren't, which is a regression from trunk I think.

With preset:

<img width="500" alt="Screen Shot 2022-04-22 at 8 53 28 am" src="https://user-images.githubusercontent.com/6458278/164564359-28314a46-ccac-495a-a217-973b01e621b9.png">

With custom:

<img width="500" alt="Screen Shot 2022-04-22 at 8 53 37 am" src="https://user-images.githubusercontent.com/6458278/164564357-92f545ae-54ef-4b82-b67e-48fc9f94cb8a.png">

ramonjd commented on packages/block-library/src/cover/transforms.js:
```diff
@@ -192,6 +195,72 @@ const transforms = {
+				const transformedColorAttributes = {
+					backgroundColor: attributes?.overlayColor,
+					gradient: attributes?.gradient,
+					style: attributes?.style
```
Do we need to check for `attributes?.style` here? If there is no style **and** custom colors in the attributes, the custom colors won't migrate. 

Reported in [this comment.](https://github.com/WordPress/gutenberg/pull/40293#issuecomment-1105836165)

I've been smoke testing the object without this check and seems to function okay so far...

```js
style: {
    ...attributes?.style,
    color: {
        background: attributes?.customOverlayColor,
        gradient: attributes?.customGradient,
        ...attributes?.style?.color,
    },
},
```

glendaviesnz replied:
Thanks will take a look - without the check for the style attribute it was leaving an empty style object serialised in the block attributes - and normally if the style attribute is empty it doesn't get serialised, but that is breaking the custom colors for some reason - will take a look and see if there is a way to not get the empty object, but for custom colors to also work, otherwise we might have to live with the empty attribute.

glendaviesnz replied:
@ramonjd I just pushed a fix for this - keeps custom colors, but still doesn't leave behind an empty style attribute if they aren't set 🤞

ramonjd replied:
Thanks @glendaviesnz LGTM

Commit:
Check for specific style attributes rather than just the top-level style object

Commit:
Apply style object update to group from transform

Commit:
Remove redundant or check

ndiego commented:
Are we still aiming to get this into 6.0? If not, I will remove from the project board. Thanks!

glendaviesnz commented:
> Are we still aiming to get this into 6.0? If not, I will remove from the project board. Thanks!

I don't think this is critical for 6.0, and because it also relies on [another fix to work](https://github.com/WordPress/gutenberg/pull/40497), I think we should defer to 6.1.

aaronrobertshaw commented:
I believe we'll also need to avoid unnecessary wrapping while transforming from the other direction i.e. from Group. 

I've created https://github.com/WordPress/gutenberg/pull/40602 to illustrate and fix this separately given I don't think it requires https://github.com/WordPress/gutenberg/pull/40497 to land.

andrewserong committed:
Move check down to style.color instead of the root style object, preserve clearing out background / gradient in transform from the Group block

andrewserong commented:
> but added an empty `style: { color: {}}` attribute if custom colors were not set, so have pushed an couple of updates to prevent this.

Thanks for adding in the fixes @glendaviesnz (and the related #40497) and for resolving the unnecessary wrapping when transformed from Group block @aaronrobertshaw! I've just given this a rebase and re-tested.

@glendaviesnz it looks like in https://github.com/WordPress/gutenberg/pull/40293/commits/796ca6866101c5c0ee39d1509a850efe2fd8ae8e the background and gradient styles are now added in to the `attributesWithoutBackgroundColors` object. Was that intentional? My thinking at the time was that we'd strip the background and gradient colors from the Group block when transforming from Group, since we're moving those colors to the parent.

I've pushed an update in https://github.com/WordPress/gutenberg/pull/40293/commits/508711d9620842c0d16ae93f95591e5669ad33b6 that moves the check down to `style.color` instead of the root `style` (so that we still preserve other `style` properties, like padding in the transform), and I've updated it to the preserve the removal of the background / gradient of the Group block:

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/166196504-1cbab09b-58af-442c-bf0b-f3f80b9a8a51.png) | ![image](https://user-images.githubusercontent.com/14988353/166196524-1f4832b6-71c2-4edd-9b99-9b706b9b1fc2.png) |

I've also added in calls to `cleanEmptyObject`, so I _think_ this still avoids the issue of serializing the empty `style: { color: {}}` object, but please let me know if I missed anything / added in a regression to your fixes! 🙇

glendaviesnz reviewed (approved):
This tested well for me now. 

✅ Transformed range of Cover blocks with preset and custom colors and gradient to Group and back and no extra wrappers added
✅ Transformed Cover with background image and it went to a Group with nested Cover block

I wonder if it is worth adding some tests for this? packages/blocks/src/api/test/factory.js has some examples of how to test transformations - maybe it would be good to start adding some block specific ones to the block folders as we modify transformations to prevent regressions - I haven't looked closely at how easy/feasible it is to mock a specific transformation though, and not a blocker for this PR, could be a follow up if you think it is feasible.

andrewserong commented:
Thanks for re-testing @glendaviesnz!

> I wonder if it is worth adding some tests for this?

That's a great point, I'll do a little digging to see how feasible it is, but it'd be good to add some in, given that we have a bit of subtle logic going on here that'd be easy to accidentally break. I'll see how I go!

andrewserong committed:
Add tests for transforms

andrewserong commented:
> packages/blocks/src/api/test/factory.js has some examples

Thanks again @glendaviesnz, that worked a treat, after a little digging, it wound up being fairly straightforward to add in tests for the transformations without having to do any mocking, I could just register the Cover and Group blocks via their metadata and then call `switchToBlockType` 👍

I also added in a test to cover the behaviour @aaronrobertshaw added in #40602.

Do you mind giving this another look when you get a chance, to confidence check the testing approach? Thanks!

aaronrobertshaw commented on packages/block-library/src/cover/test/transforms.js:
```diff
@@ -0,0 +1,272 @@
+	} );
+
+	describe( 'transform from Cover to Group', () => {
+		it( 'should transform named gradient color to Group block', () => {
```
```suggestion
		it( 'should transfer named gradient color to Group block', () => {
```

At first glance, this reads like you are transforming a color to a block. Similar sort of things for the rest of the tests within this `describe` block. Would a small tweak here make them all read a little smoother?

aaronrobertshaw reviewed (commented):
Thanks for adding the tests, especially those covering #40602 🙇 

I've given this another test run:

✅ Transforming Cover to Group works well
✅ Transforming Cover blocks with image or video backgrounds work
✅ Transforming Group to Cover works well
✅ Transforming Group block variations work e.g. Row / Stack
✅ Custom & named colors/gradients are moved appropriately between transformed blocks
✅ Unnecessary wrapping and empty attributes are avoided
✅ Unit tests are all passing

❓ I may have run into an issue when testing transforms of a Cover block with a single Group block as its child. I expected the Group block's colors to be applied to the result. When the original Cover block has a gradient background, the Group block's background color is lost. When the Cover has a flat background color, the Group's background color is honored. So I feel like this is a bit inconsistent. What do you think?


https://user-images.githubusercontent.com/60436221/166421913-24ac2fa0-bd1f-45e6-b198-6aef14d6c47a.mp4



I also have one tiny suggestion to perhaps make the test names read a fraction easier but I'll leave that to you as it is completely optional.

glendaviesnz commented:
Nice to see how easily the transform tests can be done  - a good example to apply to other blocks as they are refactored. 

Agree with Aaron's note about the slight clarification of wording. Happy to retest once you have looked at the other potential issue he identified.

andrewserong committed:
Rename transform to transfer in test names for better clarity

andrewserong committed:
Skip merging Cover background colors into Group if Group contains any bg color or gradient

andrewserong commented:
Thanks for the feedback Aaron and Glen! I've done the following:

* Renamed `transform` to `transfer` in the test descriptions
* Added an `if` statement in https://github.com/WordPress/gutenberg/pull/40293/commits/2279b98832f33acf20272ce3f1e4555906da2d3b to bail out of merging Cover background colors and gradient to the Group block, if the Group block has any kind of background or gradient color — I _think_ this gets us to a better consistency where the Group block's styling is preserved if it has any. I've added another test to try to cover this, too (or at least the happy path for it). In the editor it looks like this:

![image](https://user-images.githubusercontent.com/14988353/166610571-56b0c952-7b0d-4f6d-b3c5-fdc35aa2eb08.png)

Please let me know if you can see any issues with that approach!

aaronrobertshaw reviewed (approved):
Thanks for the speedy updates @andrewserong 🚀 

Everything looks to be working well. The new test is another welcome addition and the test wording reads much better to me now.

I think this is good to go 🚢

glendaviesnz commented:
Change tested well for me :shipit: 

I wonder if the transform tests would be worthy of a blog post?

andrewserong commented:
> I wonder if the transform tests would be worthy of a blog post?

Oh, nice idea. I might write something up!
